File: mqtt_panel/web/widget/select.js
$(function() {
    $('.widget-select').on('update', function(event, blob) {
        $(this).data('mtime', blob.mtime);
        $(this).data('value', blob.value);

        $(this).find('.value-item').addClass('d-none');
        $(this).find('.value-' + blob.value).removeClass('d-none');

        $(this).trigger('enable');
        $(this).trigger('release');
    });

    $('.widget-select').click(function() {
        if ($(this).data('enable') == false) { return; }

        if ($(this).data('pending') == true) {
            $(this).trigger('release');
            return;
        }

        let This = this;

        let buttons = [];
        $(this).find('.value-item').map(function() {
            let name = $(this).data('name');
            let ro = $(this).data('ro');
            if  (name != null && ro == false) {
                let button = {
                    icon: $(this).data('icon'),
                    color: $(this).data('color'),
                    text: $(this).data('text'),
                    confirm: $(this).data('confirm'),
                    click: function() {
                        $(This).data('pending', true);
                        $('#app').trigger('widget', {
                            id: $(This).data('id'),
                            value: name,
                        });
                    },
                };
                buttons.push(button);
            }
        });

        widget_select_modal({
            message: $(this).find('.title').text(),
            timeout: null,
            cancel: function () {
                $(This).trigger('release');
            },
            buttons: buttons,
        });
    })

    $('.widget-select').each(function() {
        widget_clickable(this);
    });
});

function widget_select_modal(blob) {
    let window = $('<div></div>')
    let close = $('<span class="modal-close material-icons">close</span>');
    let panel = $('<div></div>')
    window.append(close);
    window.append(panel);

    panel.append($('<div class="modal-message">' + blob.message + '</div>'));

    let buttons = $('<div class="modal-select"></div>');

    for (let i in blob.buttons) {
        let button = blob.buttons[i];
        let icon = '';
        if (button.icon !== undefined) {
            icon = button.icon;
        }
        let color = '';
        if (button.color !== undefined) {
            color = ' style="color:' + button.color + '"';
        }
        let html = $('<div></div>')
        html.append($('<span class="material-icons"' + color + '>' + icon + '</span>'));
        html.append($('<span' + color + '> ' + button.text + '</span>'));
        let click = function() {
            $('#modal').trigger('close');
            button.click();
        }
        if (button.confirm) {
            $(html).click(function(event) {
                event.stopPropagation();
                $('#modal').trigger('close');
                $('#modal').trigger('confirm', [{
                    message: button.confirm,
                    proceed: click,
                    cancel: blob.cancel,
                    timeout: 7000,
                }]);
            });
        } else {
            $(html).click(function(event) {
                event.stopPropagation();
                click();
            });
        }

        buttons.append(html);
    }

    $(window).click(function(event) {
        event.stopPropagation();
    });
    $(close).click(function() {
        $('#modal').trigger('cancel');
    });

    panel.append(buttons);

    $('#modal').trigger('show', [window, blob.cancel, blob.timeout]);
}

File: mqtt_panel/web/widget/select.py
from mqtt_panel.web.widget.switch import Switch


class Select(Switch):
    widget_type = 'select'

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.default = Default

    def _html_write(self, fh, ctx, state):
        color = ''
        if state.color:
            color = f' data-color="{state.color}"'
        self._write_render(fh, '''\
            <div class="value-item value-{state.name}{ctx.show}"{ctx.confirm}{color} data-icon="{state.icon}" data-text="{state.text}" data-name="{state.name}" data-ro="{state.ro}">
                <span class="material-icons"{ctx.color}>{state.icon}</span>
                <span{ctx.color}>{state.text}</span>
            </div>
        ''', {'ctx': ctx, 'state': state, 'color': color}, indent=6)


class Default:
    _map = {
        ('on', 'true'): ('toggle_on', '#52D017'),
        ('off', 'false'): ('toggle_off', 'black'),
        None: ('radio_button_unchecked', None),
    }

    @classmethod
    def _lookup(cls, *keys):
        for key in keys:
            key = key.lower()
            for k, v in cls._map.items():
                if k and key in k:
                    return v
        return cls._map[None]

    @classmethod
    def icon(cls, *key):
        return cls._lookup(*key)[0]

    @classmethod
    def color(cls, *key):
        return cls._lookup(*key)[1]
